group-coordinator/src/test/java/org/apache/kafka/coordinator/group/MockCoordinatorTimer.java

```diff
@@ -0,0 +1,144 @@
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.PriorityQueue;
+import java.util.concurrent.TimeUnit;
+
+public class MockCoordinatorTimer<T> implements CoordinatorTimer<T> {
+    public static class ScheduledTimeout<T> {
+        public final String key;
+        public final long deadlineMs;
+        public final TimeoutOperation<T> operation;
+
+        ScheduledTimeout(
+            String key,
+            long deadlineMs,
+            TimeoutOperation<T> operation
+        ) {
+            this.key = key;
+            this.deadlineMs = deadlineMs;
+            this.operation = operation;
+        }
+    }
+
+    public static class ExpiredTimeout<T> {
+        public final String key;
+        public final List<T> records;
+
+        ExpiredTimeout(
+            String key,
+            List<T> records
+        ) {
+            this.key = key;
+            this.records = records;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+
+            ExpiredTimeout<?> that = (ExpiredTimeout<?>) o;
+
+            if (!Objects.equals(key, that.key)) return false;
+            return Objects.equals(records, that.records);
+        }
+
+        @Override
+        public int hashCode() {
+            int result = key != null ? key.hashCode() : 0;
+            result = 31 * result + (records != null ? records.hashCode() : 0);
+            return result;
+        }
+    }
+
+    private final Time time;
+
+    private final Map<String, ScheduledTimeout<T>> timeoutMap = new HashMap<>();
+    private final PriorityQueue<ScheduledTimeout<T>> timeoutQueue = new PriorityQueue<>(
+        Comparator.comparingLong(entry -> entry.deadlineMs)
+    );
+
+    public MockCoordinatorTimer(Time time) {
+        this.time = time;
+    }
+
+    @Override
+    public void schedule(
+        String key,
+        long delay,
+        TimeUnit unit,
+        boolean retry,
+        TimeoutOperation<T> operation
+    ) {
+        cancel(key);
+
+        long deadlineMs = time.milliseconds() + unit.toMillis(delay);
+        ScheduledTimeout<T> timeout = new ScheduledTimeout<>(key, deadlineMs, operation);
+        timeoutQueue.add(timeout);
+        timeoutMap.put(key, timeout);
+    }
+
+    @Override
+    public void cancel(String key) {
+        ScheduledTimeout<T> timeout = timeoutMap.remove(key);
+        if (timeout != null) {
+            timeoutQueue.remove(timeout);
+        }
+    }
+
+    public boolean contains(String key) {
+        return timeoutMap.containsKey(key);
+    }
+
+    public ScheduledTimeout<T> timeout(String key) {
+        return timeoutMap.get(key);
+    }
+
+    public int size() {
+        return timeoutMap.size();
+    }
+
+    public List<ExpiredTimeout<T>> poll() {
```

Comment:
So we don't expire timeouts until we call this poll?

Reply:
Added some javadoc.